### DISCUSSION TITLE
What kind of Sensors should we support?

### DISCUSSION THREAD

**Discussion by emiltin** [rsmp-nordic/rsmp_sxl_sensors#2] [Ideas]
RSMP Nordic is going to support Sensors with a standardized SXL which we (RSMP Nordic) will maintain.

Many different types of sensors exists. Which types should we support? Our aim is not to support everything, but rather to find a sweet spot where we support 80-90% of what road authorities need, without having to spend a lot of energy on supporting very specialized equipment.

- Simple on/off sensor
- Push buttons
- Traffic counters based on video, tubes, radar
- Lidar
- etc.

**Comment by KK-KBM** [comment]
The question is rather: which way should we support sensors? Today a sensor is just an input, more os less regardless of type. Should rsmp support an object stream for a sensor (aka radar, lidar)? Counters are to some extent supported via s0201-4.

**Reply by emiltin** [reply]
> The question is rather: which way should we support sensors? Today a sensor is just an input, more os less regardless of type. Should rsmp support an object stream for a sensor (aka radar, lidar)? Counters are to some extent supported via s0201-4.

Yes we can distinguish between streaming data and aggregated data. But I think that is not so much dependent on the equipment type. The update rate you chooce when subscribing to data determines whether you're streaming data or received aggregated (or snapshot) at a regular interval.

We should try to design SXLs that are flexible enough to work with many types of VMSs and sensors. But I think it's useful to define a set of concrete types that we know we must support from the start. For example is a requirement to support lidar? Or should be postpone thinking about that and instead focus on types we know are in wide use by road authorities,. eg:

- Video sensor
- Loops
- Pneumatic tubes
- Push buttons
- Radars

The sensors typically delivers data about:

- counts
- speeds
- occupancy
- classification
- 
Basically I think we need to focus a bit, to get anywhere.

**Reply by otterdahl** [reply]
> Should rsmp support an object stream for a sensor (aka radar, lidar)?

I think we should at some point continue working (https://github.com/rsmp-nordic/rsmp_sxl_traffic_lights/issues/76) on receiving real time data from detector loops.

**Reply by rafaelcf2** [reply]
RSU (Road side unit) OBU (On Board Unit) in short range are also good usecases
some give direction vector, speed, distance

**Reply by yco-svms** [reply]
from my point of view: my notion of stream of data represents raw data, not elaborated like a video stream. I don't think there's any point in including this type of stream in RSMP exchanges, as other efficient solutions exist... (let's be lazy)

I take the principle that data from sensors is always processed data (or considered as such) and I make the following distinction: synchronous data and asynchronous data.

- _Asynchronous data_ is data that 'appears' without any temporal periodicity (a vehicle passing by, , a request to pass by a pedestrian, for example). The notion of RSMP subscription corresponds well to the "real-time" processing of this type of data; it can however be collected by periodic polling, but in this case the recovery period must be adjusted according to the expected "freshness" of the data.

- _Synchronous data_ is data that is periodically aggregated by the sensor (weather data, classified vehicle flow over the last hour, for example) ; in this case, polling is sufficient to collect it, even though it can be pushed by the sensor as soon as it is generated.

In conclusion, there is no real difference between these two types of data in terms of how they are transported by RSMP. It's more a question of application differentiation.

_PS
My initial proposal on raw data may seem abrupt. However, transporting raw data on a constant stream for remote post-processing imposes additional constraints such as: managing transport latency, the need to specify all types of raw data in the SXLs (or failing that to set up a tunnel), to have a dedicated transport mechanism in RSMP core. There are several reasons for having/wanting to deport the processing of the stream to another network node, however I think that in this case the sensor in the SXL sense becomes the post-processing unit and not the source of the stream._

**Comment by emiltin** [comment]
Sensors often have a concept of detection zones. Is that something we need to support? Or could different components be used for that? I imagine it would be nice to be able to get both data for all zones, or split by zone.

To what degree do we want the sensor to deliver raw data, vs. data aggregated or processed in various ways?

**Comment by emiltin** [comment]
some sensor types, like radars can provide many different types of data, like:
- location
- velocity
- acceleration
- ETA

**Comment by emiltin** [comment]
WG: Our initial focus will be to focus on:
- Inductive loops
- Radars
- Video sensor
- Push buttons
- Tubes

Less common use cases that we can consider later:
- Gates (eg. RFID tags, number plates, etc)
- Lidar
- Magnometers
- Location of public transport (busses, trams). eg. IMPC, GPS, RFID, often via geofencing in a central system. more related to priority than to traffic data?

 
Travel Time is another story, and can be computed with eg:
- Automatic Number Plate Recognition
- WIFI/BLuetooth
- Video

**Comment by emiltin** [comment]
WG: data types we will initially focus on:
- count
- speed
- occupancy

Other things to consider later:
- direction?
- events - stopped vehicles, peoples behavior, dropped cargo, smoke... 
- ETA?
- raw data?
- meta data: quality/accuracy
- turn movements in intersections
- 

Aggregate by:
- time period
- classification - but many different standards exists in EU
- detection zone

Streaming data:
rsmp does not currently provide a way to stream data that contain every single data point, except by sending every detection as a message, which is inefficient. we need a way to send package that contains a list of single data points (much like video or audio is commonly streamed).

GDRP and privacy:
it's an important consideration (perhaps especially for video detection?)

Data size:
Sending large amounts of data might require ways to prioritize message types in rsmp. For example if you want to send or stream a lot of traffic data.
